Revisiting the visit API (aka FastBoot™)

chancancode opened the pull request:
:arrow_right_hook: :arrow_right_hook: :arrow_right_hook: :arrow_right_hook: :arrow_right_hook: :arrow_right_hook: :arrow_right_hook: :arrow_right_hook: :arrow_right_hook: :arrow_right_hook: :arrow_right_hook: :arrow_right_hook: :arrow_right_hook: :arrow_right_hook: :arrow_right_hook: :arrow_right_hook: :arrow_right_hook: :arrow_right_hook: :arrow_right_hook: :arrow_right_hook: :arrow_right_hook: :arrow_right_hook: :arrow_right_hook: :arrow_right_hook: :arrow_right_hook: :arrow_right_hook: :arrow_right_hook:

It's time to put on some finishing touches for the `visit` API! Since this is a somewhat big patch, here is a detailed description.
# Summary

The `visit` API exposes `ApplicationInstance` as a useful primitive for many scenarios, including FastBoot™, widgets, resource discover, testing, and more.
## Background

Ember 1.12 introduced the concept of `ApplicationInstance`s. While they are very well designed and could be very useful in many cases, there are no official APIs to create them. The `visit` API helps fill this void by exposing a supported way to boot and customize instances.
## What's in the box :package:
- [x] Refactoring
- [x] New APIs
- [x] Tests
- [x] Documentation
- [x] Bonus DLC Pack
- [x] Emojis
### Refactoring
1. **A Clear Boot Phase**
   
   Previously, we don't have a very clear concept of a boot phase, especially for `ApplictionInstance`. This creates a lot of implicit timing dependencies around when certain options can be set/changed. For example, the `ember-cli-fastboot` changes the `autoboot` flag in an application initializer (i.e. in the middle of the boot phase), which created a strange fork in the boot flow.
   
   As part of this patch I did some work to solidify the boot phase. Options like `autoboot` should be set
   ~~before `boot()` is called~~ at instantiation time. What I can do on `Application` is rather limited, but I tried to make that clear in the new `ApplicationInstance` boot phase, which takes a hash of boot-time options in favor of ad-hoc methods like `overrideRouterLocation`.
2. **Asynchronous APIs**
   
   Booting an `Application` (and to a lesser extent, an `ApplicationInstance`) is an inertially asynchronous task. Naturally, we would like to express the asynchrony here with promises/async functions. However, the architecture we have predated Ember's adoption of promises, and a lot of code assumes that it is a "synchronous". Specifically, a lot of tests assumes the last call to `app.advanceReadiness()` or `app.reset()` will complete the boot process when the same runloop ends, which makes it impossible to async-ify the process.
   
   This wasn't clear to me when I started working on the code – I actually spent some time making them fully async, only to find a lot of tests failing. However, we should start embracing the asynchrony here going forward. I converted the primary (internal) APIs async, which we can start using for new code like the `visit` API. Internally, they wrap their synchronous counterparts which we will use for legacy paths.
   
   This makes things a bit annoying in the short term, but it should help us iterate towards the right direction.
3. **Globals Mode**
   
   Previously, setting the `autoboot` flag also skipped some seemingly unrelated code paths in the boot process. It turns out that most of these code paths are there for the legacy "globals mode" (where the application object acts as a "namespace" for attaching classes, i.e. `App.MyComponent = ...`).
   
   While it's fair to overload the `autoboot` flag as an opt-out for these features, they are ultimately unrelated concerns. To make the distinction clear, I added an internal `_globalsMode` flag to track these code paths. Its value is currently tied to the `autoboot` flag, but you can manually hardcode the `autoboot = true, _globalsMode = false` combo and run the tests, which should reveal that spots where we are relying on these behaviors. Since most of these features have already been deprecated, separating it from `autoboot` as a concept should help to review and remove them independently.
4. **Expansion of the `environment` concept**
   
   With HTMLBars, most of the initial render path has already been jQuery-free (which is why FastBoot works in node). The last bit I had to plumb was the `view.appendTo(rootElement)` call, which uses jQuery to resolve the `rootElement` selector into an Element.
   
   The current `visit` implement works around this by exposing the "root view" without attaching its element. This isn't ideal since we already deprecated/removed views and it requires the use of other private APIs like `appendTo` and `renderToElement`.
   
   In practice, if we were given an actual `Element` object as the `rootElement`, our renderer already knows how to append it into the DOM without jQuery. All I need to do is to tell plumb that through to the view layer so that it can skip the jQuery lookup if the user has opt-ed to not user jQuery. To do this, I expanded on the `environment` abstraction and made it instance specific through the `-environment:main` container key. I only did what I needed to do to make my scenarios work, but that should lay the ground work for the eventual "optional jQuery" world.

See also https://github.com/emberjs/ember.js/issues/11291 and https://github.com/emberjs/ember.js/pull/11289#issuecomment-105960488
### New (Public) APIs
- `Application.autoboot` flag
- `Application#boot()`
- `Application#visit()`
- `ApplicationInstance#getURL()`
- `ApplicationInstance#visit()`
### Tests

I expanded the unit tests coverage, and added some end-to-end integration tests for the a few target scenarios we intend to support.
### Documentation

You can find them in the diff.
### Bonus DLC Pack

To redeem, enter the following code into the Ember Data store: `F83J-7BD2-D16Z-3QA4`
### Emojis

:+1: :star: :point_right: :heart: :headphones: :key: :rocket: :japanese_goblin: 
## Differences from current implementation
- `Application.autoboot` should only be set before boot time (e.g. at `create` time)
- `Application#visit()` no longer rejects if a redirect occurred during the initial transition
- `Application#visit()` no longer attach a `view` property to the instance it returns
- `Application#visit()` no longer prevents the router from inserting the top-level view into the DOM (use `rootElement` to customize the insertion point as usual)
- `Application#visit()` resolves only after rendering is complete and the root view is attached to the DOM
- `Application#visit()` now takes an option hash for customizing the instance
- Uncaught errors in application/instance initializers now causes `Application#visit()` to reject instead of synchronously throws
- Removed `ApplicationView#overrideRouterLocation` (use the `location` boot option instead)
- New flags that enables a much wider range of scenarios
- Better documentation
## TODOs
- [ ] Add some node tests
- [ ] Update the `ember-cli-fastboot` implementation
- [ ] Move [`ember-cli-fastboot`-specific hacks](https://github.com/tildeio/ember-cli-fastboot/blob/master/app/initializers/fastboot.js#L12-L20) into the appropriate package; since the DOMHelper API is not public, we cannot ship until things works out of the box (see also https://github.com/krisselden/morph-range/pull/7)
- [ ] Better error reporting: differentiate between a `BootError` and a `TransitionError`: for transition errors, it is useful to still return the instance, as there we might have rendered an error page into the DOM
## Open questions
- [ ] Various TODOs in the code comments that I have no answers to
- [ ] Are we comfortable committing the the three scenarios I listed in the documentation? Do we need to add more?
- [ ] Is the `ApplicationInstance#visit` implementation correct?
- [ ] Can we consolidate `autoboot` with Ember CLI's `autorun` flag? (they do related but not identical things)
## Future extensions
- Formalize the API requirements for the fake DOM, DOMHelper, etc
- Use these API for tests: boot a new instance in each test case by calling `visit()` and `destroy()` it in teardown, instead of calling `App.reset()`. Use `ApplicationInstance#visit` to implement the `visit` test helper, etc
- Work with @mitchlloyd on Ember Islands type scenarios (see the [test](https://github.com/emberjs/ember.js/commit/b037894a2d278effb07f469d64cb68c7ec0739b8))
- Flesh out the options to enable other scenarios
- Plumb the `environment` through the rest of the view layer to make `jQuery` optional
- Remove reliance on globals mode in tests
- Add a mechanism to differentiate between the various types of redirects and common errors, so that a FastBoot server can return the right status code

cc @wycats @tomdale @stefanpenner @krisselden @mitchlloyd @chadhietala @joshvfleming :pray:

chancancode committed:
Untangle autoboot and globalsMode code paths

Conceptually, these are two orthgonal things – autoboot happens to be
the opt-in for disabling support for globals mode.

After this refactor, autoboot is literally just calling `boot()` at
DOM Ready and *then* creating a instance automatically for you.
Conversely, you mustcall `boot()` yourself if you disable autoboot).

Most of the globals mode featue are already deprecated in 1.x.
Separating the two concepts allow us to review and remove those
feaatures independently.

Locally overriding the flags to use the `(autoboot: true, globalsMode:
false)` combo and running the tests would reveal all the places we are
relying on them internally.

chancancode committed:
Fully async-ify the boot process

chancancode committed:
Make App.visit() boot the app if needed

chancancode committed:
Finalize namespace before booting instances

chancancode committed:
Refactor application (instance) boot paths

This is a dialed back attempt at making the `ApplicationInstance`
boot process fully async. I wrote all the code to make that happen,
only to find out that a lot of tests implicitly depend on the process
being synchronous. So, I did the following instead:

1. Create an async version of `boot` on both `Application` and on
   `ApplicationInstance`. Going forward, new APIs (`visit`, etc)
   should use the async version.

2. Keep an internal `_bootSync` method on both classes and use them
   in places where we rely on the process being synchronous.

In the short term, having to support both synchronous and
asynchronous might make things look a bit more complicated, but it
will help us iterate towards where we want to be.

Separately, this commit also solidified the concept of a `boot`
phase on `ApplicationInstances`. Instead of overriding things on
the instance in an ad-hoc manner and implicitly rely on the timing
of these calls (e.g. `overrideRouterLocation` must be called before
the router is instantiated/cached), they have now been consolidated
into options that you pass into the `boot()` method.

Having an explicit boot phase makes it clear what can and cannot
happen at runtime.

(TODO: move the "start routing" logic into the boot phase)

chancancode committed:
Properly handle redirects and aborts

chancancode committed:
Add a FastBoot test to help design the API

chancancode committed:
Ember Islands... it Just Works™ o_0

chancancode committed:
Docs; tighten the public API; resource discovery

chadhietala commented on packages/ember-application/lib/system/application.js:
```diff
@@ -456,50 +577,74 @@ var Application = Namespace.extend(RegistryProxy, {
-    Initialize the application. This happens automatically.
+    By default, this method is called automatically on "DOM ready"; however, if autoboot
+    is disabled, you would need to call this method yourself. Calling this method before
+    "DOM ready" is not recommended.
```
I wouldn't say it's not recommended but rather there are caveats. The caveats being that your application cannot rely on DOM readiness.

chancancode replied:
I actually had your use case in mind when I refactored that path, but I think the plot was lost somewhere :P I'll re work this part a bit more to make that more

tomdale replied:
What is @chadhietala's use case?

chancancode replied:
They had a streaming (as in HTTP chucked response) setup that keeps the body tag open for sending data (<script> ...json... </script>), which causes the DOM ready to fire very late, which delayed ember boot for no reason.

The actual caveats are rather nuanced:

1. when you kick off the boot, you must already have everything you need for your app, e.g. if your initializer and a route uses some global Google Analytics or Intercom thing, you must not boot it before they are loaded

2. When Ember eventually inserts the top level view (`didCreateRootView`), the rootElement must already be on the page

It is actually not that easy to determine these two things, and for a normal set up the gain is very little, so I think it's too nuanced to describe and recommend publicly in the docs.

However, for @chadhietala's use case, it is not that hard to determine "would have been dom ready if not for the streaming setup" point – you can just insert a synchronous script tag right before you start streaming data, so I was going to make sure that there is a thing you can call to tell Ember this is the "effective" DOM ready. Perhaps I'll just rename _autoBoot back to DOM ready or something

(I don't think we need to document this)

tomdale commented on packages/ember-application/lib/system/application-instance.js:
```diff
@@ -108,26 +118,103 @@ let ApplicationInstance = EmberObject.extend(RegistryProxy, ContainerProxy, {
-  router: computed(function() {
-    return this.lookup('router:main');
-  }).readOnly(),
+  _bootOptions: null,
```
These should be documented, even if private.

tomdale commented on packages/ember-application/lib/system/application-instance.js:
```diff
@@ -79,7 +87,9 @@ let ApplicationInstance = EmberObject.extend(RegistryProxy, ContainerProxy, {
     var application = get(this, 'application');
 
-    set(this, 'rootElement', get(application, 'rootElement'));
+    if (!isEnabled('ember-application-visit')) {
```
Why is this feature flagged? Application instances aren't exposed via public API if the flag is disabled, right?

chancancode replied:
They can't be created manually via public API, but they are used by Ember. Basically this is just trying to be carefully are preserve the /exact/ point where this was happening (I moved this exact call down into the boot method), I will be happy to unify them below if that's fine with you

tomdale commented on packages/ember-application/lib/system/application-instance.js:
```diff
@@ -108,26 +118,103 @@ let ApplicationInstance = EmberObject.extend(RegistryProxy, ContainerProxy, {
+    let defer = new Ember.RSVP.defer();
+    let promise = this._bootPromise = defer.promise;
+
+    try {
```
Why use a deferred here? Isn't this the re-implementing the same semantics as if you wrapped it in a `Promise`?

chancancode replied:
Good question! it doesn't look like it's necessary anymore

stefanpenner replied:
yup, looks like the current refactoring would be better expressed via promise constructo

```js
return new Ember.RSVP.Promise((resolve) => resolve(this._bootSync(options)));
```

tomdale commented on packages/ember-application/lib/system/application-instance.js:
```diff
@@ -108,26 +118,103 @@ let ApplicationInstance = EmberObject.extend(RegistryProxy, ContainerProxy, {
+    We would like new code (like the `visit` API) to stop making this assumption,
+    so we created the asynchronous version above that returns a promise. But until
+    we have migrated all the code, we would have to expose this method for use
+    *internall* in places where we need to boot an instance synchronously.
```
s/internall/internally/

tomdale commented on packages/ember-application/lib/system/application-instance.js:
```diff
@@ -108,26 +118,103 @@ let ApplicationInstance = EmberObject.extend(RegistryProxy, ContainerProxy, {
-    var location = options && options.location;
-    var router = get(this, 'router');
+  _bootSync(options) {
+    if (this._booted) { return this; }
```
Are there valid use cases for absorbing a double call to `_bootSync`? I can imagine that if the second call has different options (that are thus ignored) and there's no error, tracking that bug down will be extremely annoying.

chancancode replied:
Hm, I was just mirroring the semantics of the async path. Maybe calling boot **with options**, on an already booted instance should result in a rejection? Or would that just be a normal assertion (exception)?

tomdale commented on packages/ember-application/lib/system/application-instance.js:
```diff
@@ -108,26 +118,103 @@ let ApplicationInstance = EmberObject.extend(RegistryProxy, ContainerProxy, {
+      let registry = this.__registry__;
+      let environment = options.toEnvironment();
+
+      set(this, '_environment', environment);
```
Does this need to be saved on both the application instance instance (lulz) and registered in the registry? Seems like just the registry is enough?

chancancode replied:
You are right! I was using it for something else but I alter removed that code, :scissors:

tomdale commented on packages/ember-application/lib/system/application-instance.js:
```diff
@@ -212,6 +298,255 @@ let ApplicationInstance = EmberObject.extend(RegistryProxy, ContainerProxy, {
+      @default auto-detected
+      @private
+    */
+    this.jQuery = jQuery;
```
It seems like this isn't overridable via `options` and always uses the imported jQuery?

tomdale replied:
Ahh, I see it down below now. Seems strange not to initialize it to the `options` value until the end?

chancancode replied:
Yeah, this was a bit awkward. The `browser` and `render` options are "connivance options" or "combo options" that implies a bunch of other (internal) flags. So what really needs to happen is:

1. If the user supplied one of the combo flags, do the default for that option (browser implies `jQuery = null`)
2. But if, for some reason, they really really want to override it and supplied a `jQuery` (or location, etc) in addition to the combo flag, let them do it
3. Then finally, if they didn't specify either, use the default (system `jQuery`)

..which is why it ended up looking like this

tomdale commented on packages/ember-application/lib/system/application-instance.js:
```diff
@@ -212,6 +298,255 @@ let ApplicationInstance = EmberObject.extend(RegistryProxy, ContainerProxy, {
+    /**
+      Provide a specific instance of jQuery. This is useful in conjunction with
+      the `document` option, as it allows you to use a copy of `jQuery` that is
+      appropiately bound to the foriegn `document` (e.g. a jsdom).
```
s/appropiately/appropriately/
s/foriegn/foreign/

tomdale commented on packages/ember-application/lib/system/application-instance.js:
```diff
@@ -212,6 +298,255 @@ let ApplicationInstance = EmberObject.extend(RegistryProxy, ContainerProxy, {
+      @default auto-detected
+      @private
+    */
+    this.interactive = environment.hasDOM;
```
Is this the same as `destinedForDOM`?

chancancode replied:
This is `destinedForDOM` AND `please set up event delegation` (event dispatcher).. it wanted to consolidate the two into a single thing since it doesn't seem to make sense to have one but not the other, thoughts?

tomdale commented on packages/ember-application/lib/system/application-instance.js:
```diff
@@ -212,6 +298,255 @@ let ApplicationInstance = EmberObject.extend(RegistryProxy, ContainerProxy, {
+      and interactive features. Specifically:
+
+      * It does not use `jQuery` to append the root view; the `rootElement`
+        (either specified as a subsequent option or on the applicatio itself)
```
s/applicatio/application/

tomdale commented on packages/ember-application/lib/system/application-instance.js:
```diff
@@ -212,6 +298,255 @@ let ApplicationInstance = EmberObject.extend(RegistryProxy, ContainerProxy, {
+
+      * It sets the `location` option to `"none"`. (If you would like to use
+        the location adapter specified in the app's router instead, you can also
+        specify `{ location: null }` to specifically opt-out.)
```
:+1: This cleanup is great.

tomdale commented on packages/ember-application/lib/system/application-instance.js:
```diff
@@ -212,6 +298,255 @@ let ApplicationInstance = EmberObject.extend(RegistryProxy, ContainerProxy, {
+      @public
+    */
+    if (options.render !== undefined) {
+      this.render = !!options.render;
```
I wonder if these flags should be renamed so it's more obvious that they're booleans. `render` has the problem of typically being used as a verb.

Maybe `isInteractive` and `shouldRender` instead?

chancancode replied:
seems good!

tomdale commented on packages/ember-application/lib/system/application-instance.js:
```diff
@@ -212,6 +298,255 @@ let ApplicationInstance = EmberObject.extend(RegistryProxy, ContainerProxy, {
+
+  BootOptions.prototype.toEnvironment = function() {
+    let env = assign({}, environment);
+    env.hasDOM = this.browser;
```
Like above, using `browser` (a noun) to mean a Boolean is very confusing to a reader unfamiliar with the code. I think `isBrowser` is a better choice.

chancancode replied:
seems good!

tomdale commented on packages/ember-application/lib/system/application.js:
```diff
@@ -306,6 +306,52 @@ var Application = Namespace.extend(RegistryProxy, {
 
+  /**
+    Whether the application should be configured for the legacy "globals mode".
+    Under this mode, the Application object serves as a gobal namespace for all
```
s/gobal/global/

tomdale commented on packages/ember-application/lib/system/application.js:
````diff
@@ -306,6 +306,52 @@ var Application = Namespace.extend(RegistryProxy, {
+    });
+    ```
+
+    This flag also exposes other internal APIs that assumes the existance of
````
s/existance/existence/

tomdale commented on packages/ember-application/lib/system/application.js:
```diff
@@ -306,6 +306,52 @@ var Application = Namespace.extend(RegistryProxy, {
+
+    This option is currently not configurable, its value is derived from
+    the `autoboot` flag – disabling `autoboot` also implies opting-out of
+    globals mode support, although they are ultimately orthgonal concerns.
```
s/orthgonal/orthogonal/

tomdale commented on packages/ember-application/lib/system/application.js:
```diff
@@ -306,6 +306,52 @@ var Application = Namespace.extend(RegistryProxy, {
+    globals mode support, although they are ultimately orthgonal concerns.
+
+    Most of the global modes features are already deprecated in 1.x. The
+    existance of this flag is to untangle the globals mode code paths from
```
s/existence/existance/

tomdale commented on packages/ember-application/lib/system/application.js:
```diff
@@ -318,24 +364,25 @@ var Application = Namespace.extend(RegistryProxy, {
-      this.waitForDOMReady();
+      // Force-assign these flags to their default values when the feature is
+      // disabled, this ensures we can rely on their values in other paths.
+      this.autoboot = true;
```
Should rewrite to `this.autoboot = this._globalsMode = true` for visual symmetry with above branch of the conditional.

chancancode replied:
seems good!

tomdale commented on packages/ember-application/lib/system/application.js:
```diff
@@ -353,56 +400,130 @@ var Application = Namespace.extend(RegistryProxy, {
-    // For the default instance only, set the view registry to the global
-    // Ember.View.views hash for backwards-compatibility.
-    EmberView.views = instance.lookup('-view-registry:main');
+    This is orthgonal to autoboot: the deprecated instance needs to
```
s/orthgonal/orthogonal/

tomdale commented on packages/ember-application/lib/system/application.js:
````diff
@@ -353,56 +400,130 @@ var Application = Namespace.extend(RegistryProxy, {
+    }
+    ```
+
+    Unfortunately, because we need to participate in the "synchronous" boot
````
First sentence is missing a dependent clause.

tomdale commented on packages/ember-application/lib/system/application.js:
```diff
@@ -456,50 +577,74 @@ var Application = Namespace.extend(RegistryProxy, {
+    try {
+      this._bootSync();
+    } catch(_) {
+      // Ignore th error: in the asynchronous boot path, the error is already reflected
```
iScare. If `_bootSync()` wraps everything in a Promise, then it won't raise an exception. However, if an exception _is_ raised somewhere there, this will swallow it completely.

Put another way: if the Promises is responsible for handling the error, this path will never be triggered. If this path *is* triggered, then by definition the error was not reflected in the promise rejection.

chancancode replied:
This is :cry:

Basically that's part of the reason why _bootSync uses `defer` instead of wrapping it in a promise. Code paths that treats boot as synchronous (now calls `_bootSync`) *might* expect exceptions, so `_bootSync` as an API/method needs to throw. On the other hand, the async API wants to reify this as a promise rejection.

So `_bootSync` and `boot` works like this:

```javascript
boot() {
  // Async!
  try {
    _bootSync();
  } catch(_) {
    // Nope!
  }

  return promiseCreatedByBootSync;
}

_bootSync() {
    var defer = new Defer();

    try {
      doStuff();
      defer.resolve(this);
    } catch(e) {
      defer.reject(e);
      throw e; // for existing consumers
    }
}
```

chancancode replied:
I would be happy to take suggestions on making this better, but having to support both sync and async is hard :/

stefanpenner replied:
why not just just have bootSync return the promise it somewhat ironically sets as a flag.

tomdale commented on packages/ember-application/lib/system/application.js:
```diff
@@ -456,50 +577,74 @@ var Application = Namespace.extend(RegistryProxy, {
+    We would like new code (like the `visit` API) to stop making this assumption,
+    so we created the asynchronous version above that returns a promise. But until
+    we have migrated all the code, we would have to expose this method for use
+    *internall* in places where we need to boot an app "synchronously".
```
s/internall/internally/

tomdale commented on packages/ember-application/lib/system/application.js:
```diff
@@ -570,24 +715,29 @@ var Application = Namespace.extend(RegistryProxy, {
       run(instance, 'destroy');
-
-      run.schedule('actions', this, 'domReady', this.buildDefaultInstance());
+      this._buildDeprecatedInstance();
```
Any teardown we need to do before we replace the old instance with the new?

chancancode replied:
the line before calls destroy on it, I think that is all :D

tomdale commented on packages/ember-application/lib/system/application.js:
```diff
@@ -645,24 +795,58 @@ var Application = Namespace.extend(RegistryProxy, {
-
+    try {
+      // TODO: Is this still needed for _globalsMode = false?
       if (!Ember.testing) {
```
I don't remember offhand where `BOOTED` is used, but I'm pretty sure we use the container key to generate `toString()` names of objects in Ember CLI across the board and so could skip `Ember.Namespace.processAll()`.

chancancode replied:
I *think* the two are related, from what I saw in tests `BOOTED` basically means "did we process the entire namespace"? I'm.. 70% sure :P

tomdale commented on packages/ember-application/lib/system/application.js:
```diff
@@ -793,35 +978,198 @@ Application.reopenClass({
-        instance.didCreateRootView = function(view) {
-          instance.view = view;
-          res(instance);
+      When `autoboot` is disabled, calling `visit` would first cause boot the
```
I think the subject and verb are switched here? "would first cause the application to boot" instead?

tomdale commented on packages/ember-application/lib/system/application.js:
````diff
@@ -793,35 +978,198 @@ Application.reopenClass({
+      MyApp.visit("/", { location: "none", rootElement: "#container" });
+      ```
+
+      All options are optional.
````
![giphy](https://cloud.githubusercontent.com/assets/90888/10120261/a0cd253c-6467-11e5-9a94-2cefa823db43.gif)

chancancode replied:
:stuck_out_tongue: 

This is an artifact of the way I documented them using the "pseudo" `BootOptions` class, I wanted to make clear you do not have to set every property that's present on that documentation page, but on second thoughts the `@default` should already take care of that (and if not, the fact that this example doesn't set other things should imply that).

So I'll rm

tomdale commented on packages/ember-application/lib/system/application.js:
```diff
@@ -793,35 +978,198 @@ Application.reopenClass({
+      correctly today, largely due to Ember's jQuery dependency.
+
+      Currently, there are three officially supported scenarios/configurations.
+      Usages outside of these scenarios are not guarenteed to work, but please
```
s/guarenteed/guaranteed/

tomdale commented:
> Internally, they wrap their synchronous counterparts which we will use
> for legacy paths.  This makes things a bit annoying in the short term,
> but it should help us iterate towards the right direction.

Can we start deprecating the synchronous APIs with an eye towards
removing them in Ember 3.0?  You mentioned specifically
`advanceReadiness()` and `reset()` as the major culprits of
backwards-incompatibility when making these synchronous. My hunch is
that those are largely used by tests and rarely used in real
applications.

Testing infrastructure is the last major thing relying on `reset()`, and
once we move over to using the `visit()` API, and I don't think there
are many remaining reasons to use either of these APIs. (The community
has started moving away from the pattern of using `advanceReadiness` et
al. in initializers, thank goodness.)

> Globals Mode

For the objects created and installed on the namespace object, I wonder
if we can get creative with ES5 getters to issue a deprecation warning
if these are ever accessed. Hopefully this should be a non-issue for
anyone that has migrated to Ember CLI.

> With HTMLBars, most of the initial render path has already been
> jQuery-free (which is why FastBoot works in node). The last bit I had
> to plumb was the view.appendTo(rootElement) call, which uses jQuery to
> resolve the rootElement selector into an Element.

:+1:


> All I need to do is to tell plumb that through to the view layer so
> that it can skip the jQuery lookup if the user has opt-ed to not user
> jQuery. To do this, I expanded on the environment abstraction and made
> it instance specific through the -environment:main container key.

Why does it need to be instance-specific? Isn't the answer to the
question "Did the user opt-out of jQuery" app-wide? How could one
instance have jQuery but not another?

> Application.autoboot should only be set before boot time (e.g. at create time)

As I mentioned to you over IM, I'm persuaded by your argument that
changing boot parameters mid-boot (via an initializer, for example, as
we do today) is error-prone and introduces ambiguity into the boot code.
Overall, I'm :+1: on this change.

However, from experience, requiring a property to be set at
instantiation time can significantly reduce the flexibility of the
system. The most recent example of this that I can recall is when we
were doing the initial `visit()` implementation and we had to override
the location of the router. Previously, the router's `location` had to
be set at `create()` time, which led to a lot of contortions in the code
just to work around something that would have been straightforward if it
could have been provided after instantiation.

The more I think about this, the more I think the ability to configure
the application after creation remains important. You just can't always
be there at creation time when you need to be. I understand why it needs
to be this way now (to synchronously support the old globals mode), but
let's think about if there's a way to loosen this restriction in the
future once globals mode is removed.

> Application#visit() now takes an option hash for customizing the instance

Applying the options hash as properties on the instance is scary to me
since not namespacing stuff like this has bit us in the past, when we
needed to add new options that were *not* instance properties. I'd
prefer that the hash of instance properties be namespaced in the options
hash:

```js
app.visit('url', {
  instanceProps: {
    makeInstanceGreatAgain: true
  }
});
```

> Move ember-cli-fastboot-specific hacks into the appropriate package;
> since the DOMHelper API is not public, we cannot ship until things
> works out of the box (see also krisselden/morph-range#7)

What does this mean?

> Better error reporting: differentiate between a BootError and a
> TransitionError: for transition errors, it is useful to still return
> the instance, as there we might have rendered an error page into the
> DOM

:+1:. In general, debugging a failing FastBoot app is extremely painful
at the moment. Stack traces are borderline useless. Out of scope for
this discussion, but I'd love to discuss a strategy for making debugging
apps in FastBoot more pleasant.

> Can we consolidate autoboot with Ember CLI's autorun flag? (they do
> related but not identical things)

That would be great. Let's eliminate all of the hacky ways of
controlling app boot and make it as explicit and unified as possible.
Where is this code even defined? I can't find whatever handles it in
either emberjs/ember.js or ember-cli/ember-cli.

> Formalize the API requirements for the fake DOM, DOMHelper, etc

Can you expand on this a bit?

chancancode commented:
>> Move ember-cli-fastboot-specific hacks into the appropriate package;
since the DOMHelper API is not public, we cannot ship until things
works out of the box (see also krisselden/morph-range#7)

> What does this mean?

Basically, the docs are written to says "you can run this in node, this is how you do it", except when you actually try to do it without using `ember-cli-fastboot`, it doesn't work without the two or three minor patches we added in fastboot addon. However, we cannot advertise those patches publicly, because they are patching private APIs (DOM helpers), and we also cannot say this feature only works with `ember-cli-fastboot`.

So basically, we need to figure out why those hacks were necessary and move it into one of the official places (inside Ember, dom helper, or morph range, etc), whatever that takes. The feature should work as advertised out of the box with the use cases described in docs.

chancancode commented:
>> Internally, they wrap their synchronous counterparts which we will use
for legacy paths. This makes things a bit annoying in the short term,
but it should help us iterate towards the right direction.

> Can we start deprecating the synchronous APIs with an eye towards
removing them in Ember 3.0? You mentioned specifically
advanceReadiness() and reset() as the major culprits of
backwards-incompatibility when making these synchronous. My hunch is
that those are largely used by tests and rarely used in real
applications.
> 
> Testing infrastructure is the last major thing relying on reset(), and
once we move over to using the visit() API, and I don't think there
are many remaining reasons to use either of these APIs. (The community
has started moving away from the pattern of using advanceReadiness et
al. in initializers, thank goodness.)

We should mark it somewhere to remind ourselves, but we can't do it yet because we can't deprecate features without a full transition story. I think we are probably at least one release out from that, we'd need:

* A promise based initializer API? (so we just RSVP.all([all, the, initializers]))
* Flesh out the instance-based testing story; I am just "hypothesizing" that it will replace `reset` etc, but haven't actually tried it in real apps :P we also need to update the test helpers, docs etc

chancancode commented:
> The more I think about this, the more I think the ability to configure
the application after creation remains important. You just can't always
be there at creation time when you need to be. I understand why it needs
to be this way now (to synchronously support the old globals mode), but
let's think about if there's a way to loosen this restriction in the
future once globals mode is removed.

I think the way to go is to treat Application is a blueprint ("prototype"!) and move all the interesting configurations into instances – they inherit whatever from the application, but if you want to customize it, you should do it when booting the instance. (Like how the `rootElement` and `location` is overridable at instance boot time now.)

Which is why...

> Why does it need to be instance-specific? Isn't the answer to the
question "Did the user opt-out of jQuery" app-wide? How could one
instance have jQuery but not another?

We have the auto-detected default, but it's one of the options you can override at instance boot time. The technical reason why this is the case is that when we get to the point where you can boot multiple copies of the app instance in full interactive mode on different `Document` objects, each app instance would need its own copy of jQuery that is bound to the right document. (Maybe you are doing something servers-side with jsdom?)

kratiahuja commented on packages/ember-application/lib/system/application.js:
```diff
@@ -318,24 +364,25 @@ var Application = Namespace.extend(RegistryProxy, {
+    }
+
+    if (this.autoboot) {
+      this._scheduleAutoBoot();
```
If I understand the flow for boot() correctly, out of the box we are still assuming an app will wait for DOM to be ready before booting. 

If an application wants to use the public boot() function to not wait for DOM ready, it would still need to touch the constructor to call boot():
init() {
  this._super(...arguments);
  this.boot();
}

Is that correct?

My two cents:
If the above is correct, this still seems a bit hacky to me to touch the constructor to call boot(). Should we just expose a flag that determines if the application wants to wait for DOM ready or not? By default the flag is set to true.
If the flag is set to false, the constructor can call _bootSync(). 

Also we could assume the flag will only be used when the autoboot is set to true.  So probably we could update this to:
if (this.autoboot) {
  if (this.waitForDomReady) {
    // wait for DOM to be ready before booting
    this._scheduleAutoBoot();
  } else {
    // boot without waiting for DOM ready.
    this._bootSync();
  }
}

tomdale replied:
@kratiahuja I'm not sure if needing to boot the app on `init()` is common-enough of a use case that we need to privilege it with more public API.

kratiahuja replied:
Sure sounds reasonable.

stefanpenner commented on packages/ember-application/lib/system/application-instance.js:
```diff
@@ -79,7 +87,9 @@ let ApplicationInstance = EmberObject.extend(RegistryProxy, ContainerProxy, {
 
-    set(this, 'rootElement', get(application, 'rootElement'));
+    if (!isEnabled('ember-application-visit')) {
+      set(this, 'rootElement', get(application, 'rootElement'));
```
rootElement isn't observable, we should just set `this.rootElement`

chancancode replied:
I only did this to preserve the /exact/ behavior without the flag to de-risk (otherwise I could just move this into `boot` too), but I will change the feature flagged version

stefanpenner commented on packages/ember-application/lib/system/application-instance.js:
```diff
@@ -108,26 +118,103 @@ let ApplicationInstance = EmberObject.extend(RegistryProxy, ContainerProxy, {
+      });
+
+      if (options.rootElement) {
+        set(this, 'rootElement', options.rootElement);
```
i don't believe rootElement is observable, we should just use a propertySet.

stefanpenner commented on packages/ember-application/lib/system/application-instance.js:
```diff
@@ -212,6 +298,255 @@ let ApplicationInstance = EmberObject.extend(RegistryProxy, ContainerProxy, {
+        // Resolve only after rendering is complete
+        return new Ember.RSVP.Promise((resolve) => {
+          // TODO: why is this necessary? Shouldn't 'actions' queue be enough?
+          // Also, aren't proimses supposed to be async anyway?
```
they are async, but relative to the run-loop. Which is essentially our "world".

Im curious what this is trying to escape?

chancancode replied:
The tests were failing because this resolves before the rendering occurred (i.e. when this resolves, the document doesn't have the root view in it), and doing this fixes it. Is there something better I can do here? (I tried `run.once("actions", ...)` or "afterRender" etc and the doesn't seem to work)

chancancode replied:
(But because this implementation is doing a correct thing, it is also not super urgent/blocker for merge)

stefanpenner commented on packages/ember-application/lib/system/application-instance.js:
```diff
@@ -212,6 +298,255 @@ let ApplicationInstance = EmberObject.extend(RegistryProxy, ContainerProxy, {
+    @namespace @Ember.ApplicationInstance
+    @public
+  */
+  BootOptions = function BootOptions(options) {
```
`BootOptions(options = {})`

stefanpenner commented on packages/ember-application/lib/system/application.js:
```diff
@@ -353,56 +400,130 @@ var Application = Namespace.extend(RegistryProxy, {
-    return ApplicationInstance.create({
-      application: this
-    });
+  buildInstance(options) {
```
`options = {}`

stefanpenner commented on packages/ember-application/lib/system/application.js:
```diff
@@ -456,50 +577,74 @@ var Application = Namespace.extend(RegistryProxy, {
+    // boot promise exists for book-keeping purposes: if anything went wrong in
+    // the boot process, we need to store the error as a rejection on the boot
+    // promise so that a future caller of `boot()` can tell what failed.
+    let defer = this._bootResolver = new Ember.RSVP.defer();
```
this seems like it could just be re-written using the promise constructor form.

chancancode replied:
This is the kind of thing I want people to not accidentally do re above :P (if I were you I would be tempted to think this is ugly and do that refactor too).

We actually cannot :( (I also didn't introduce this, `_bootPromise` and `_bootResolver` were from original patch, they just didn't correctly reject on errors.) The main reason is that this is really a function that's broken into two halves (the work it's doing here, plus the continuation at `didBecomeReady`). If you follow that code you will see that the fulfillment is in the other half, so we actually need to save of the `resolve`/`reject`somewhere.

(And unfortunately, not using the promise constructor forces us to try/catch manually, which is the bug I'm trying to fix vs the implementation on master.)

stefanpenner commented on packages/ember-application/lib/system/application.js:
```diff
@@ -456,50 +577,74 @@ var Application = Namespace.extend(RegistryProxy, {
+    @private
+  */
+  _bootSync() {
+    if (this._booted) { return; }
```
may as well just return the promise

stefanpenner commented on packages/ember-application/lib/system/application.js:
```diff
@@ -456,50 +577,74 @@ var Application = Namespace.extend(RegistryProxy, {
-    this.advanceReadiness();
+    @private
+  */
+  _bootSync() {
```
although part of the boot is sync, that is a detail. I think this should just be `_boot` and return and memoize the promise.

chancancode replied:
The "detail" is important and is the primary reason why this method exits, though.

I wrote about it in the commit message: https://github.com/chancancode/ember.js/commit/1a423865c464847539a7f9535a49a2983d626fb4

Basically, I assumed we can just make boot async (the current code on master actually assumes that: the original `ember-application-visit` patch originally introduced `_bootPromise`), so I just went ahead and made all the things proper-async, then lost a ton of time figuring out why tests a failing and eventually realized that we actually **depend** on it being sync (in our tests, at least). See the comment above for the rest of the story.

So, there are definitely other ways to structure this, but my primary goal here is to make that implicit dependency clear and make sure others won't 1) waste time 2) have to understand the constraints before changing the semantics here (i.e .they won't change it by accident).

If you think that these are not that important, or the current implementation didn't quite achieve what I said above anyway (quite possible), then I'm happy to roll it back.

stefanpenner replied:
so basically this is a stop-gap until we fix more of the boot process? If so thats fine, as long as we eventually finish

chancancode replied:
It is a stop-gap, but we probably won't be able to completely remove it until 3.0; we should start fixing our tests and somehow deprecate them though.

Basically, we have tests that looks like this:

```javascript
test('...', function() {
  var App = ...;

  run(App, 'advanceReadiness');

  nowSynchronouslyDoStuffSinceAppIsFullyBooted();
});
```

or 

```javascript
test('...', function() {
  var App = ...;

  // ...

  run(App, 'reset');

  nowSynchronouslyDoStuffSinceAppIsFullyBooted();
});
```

(I should probably edit the comment to not say "most of them are already deprecated", turns out we still support globals mode fully?)

rwjblue replied:
> we should start fixing our tests 

100% agreed.  Can you make an issue summarizing the problem, and suggest a mechanism by which we can track down tests that rely on sync behavior?  (I think that just changing this method a tiny bit would make those tests noisey enough to track down and fix...)

> turns out we still support globals mode fully?

Not only do we still support globals mode fully, but it is the only one supported by emberjs/ember.js out of the box (aka without ember-cli + modules resolver). :trollface:

chancancode replied:
> suggest a mechanism by which we can track down tests that rely on sync behavior

I am glad you asked :wink:

I added the `_globalsMode` flag for precisely this reason. If you force `autoboot = true; _globalsMode = false` (which you can't set publicly, see `Application#init`), you will see some failing tests :smile:

chancancode replied:
Actually, I might have lied :stuck_out_tongue: `_globalsMode` is related (`__container__`, etc) but not the same. I suppose you are right, we basically need to rollback to my originally implementation locally (make `boot`, `reset` etc fully async), then run the tests.

chancancode replied:
But for example, simply changing the `_bootSync` call in `reset` to `boot` would already show you some of those failures.

rwjblue replied:
Ya, makes sense.  My point is mostly that we can defer the fixing of the tests to a future time, but we should capture your thoughts in an issue to allow that grunt work to be done without the massive overhead of fully grokking boot process.

stefanpenner commented on packages/ember-application/tests/system/initializers_test.js:
```diff
@@ -34,6 +34,42 @@ QUnit.test('initializers require proper \'name\' and \'initialize\' properties',
+    });
+
+    try {
+      app.boot().then(
```
qunit supports promises, you can just `return app.boot()` or `return.app.boot().catch(...)` for a more concise test.

stefanpenner commented on packages/ember-application/tests/system/visit_test.js:
```diff
@@ -30,59 +73,347 @@ if (isEnabled('ember-application-visit')) {
+    });
+
+    // Continue after 500ms
+    setTimeout(function() {
```
can we use promises, instead of start/stop? This will simplify stuff.

Maybe adding the following helper, would cleanup stuff:

```js
function delay(time) {
  return new Ember.RSVP.Promise((resolve) => Ember.run.later(resolve, 500));
}
```

stefanpenner commented on packages/ember-application/tests/system/visit_test.js:
```diff
@@ -30,59 +73,347 @@ if (isEnabled('ember-application-visit')) {
+          }
+        );
+      });
+    }, 500);
```
500ms feels like a very long time for a test

chancancode replied:
I inherited this from the existing code, so I don't feel comfortable making the call to change it; but you are welcomed to tell me what to do ;)

stefanpenner replied:
well it seems like we really just want to listen for the same event apps autoboot on, then timeout(0) to assert the app didn't boot)

stefanpenner commented on packages/ember-application/tests/system/visit_test.js:
```diff
@@ -30,59 +73,347 @@ if (isEnabled('ember-application-visit')) {
+        }
+      });
+
+      App.visit('/').then(
```
lets return here, and not use start/stop

stefanpenner commented on packages/ember-application/tests/system/visit_test.js:
```diff
@@ -30,59 +73,347 @@ if (isEnabled('ember-application-visit')) {
+        }
+      });
+
+      App.boot().then(() => {
```
^

stefanpenner commented on packages/ember-application/tests/system/visit_test.js:
```diff
@@ -30,59 +73,347 @@ if (isEnabled('ember-application-visit')) {
+        }
+      });
+
+      App.visit('/').then(
```
^

stefanpenner commented on packages/ember-application/tests/system/visit_test.js:
```diff
@@ -30,59 +73,347 @@ if (isEnabled('ember-application-visit')) {
+        }
+      });
+
+      App.visit('/').then(
```
^

stefanpenner commented on packages/ember-application/tests/system/visit_test.js:
```diff
@@ -30,59 +73,347 @@ if (isEnabled('ember-application-visit')) {
+        }
+      }));
+
+      App.visit('/a').then(
```
^

stefanpenner commented on packages/ember-application/tests/system/visit_test.js:
```diff
@@ -30,59 +73,347 @@ if (isEnabled('ember-application-visit')) {
         }
+      }));
+
+      App.visit('/a').then(
```
^

stefanpenner commented on packages/ember-application/tests/system/visit_test.js:
```diff
@@ -30,59 +73,347 @@ if (isEnabled('ember-application-visit')) {
+    QUnit.expect(2);
+    QUnit.stop();
+
+    run(function() {
```
Are we actually intentionally creating and visiting in the same run loop? This does not seem correct, or a natural thing to happen.

stefanpenner commented on packages/ember-application/tests/system/visit_test.js:
```diff
@@ -30,59 +73,347 @@ if (isEnabled('ember-application-visit')) {
+    QUnit.expect(8);
+    QUnit.stop();
+
+    run(function() {
```
^

stefanpenner commented on packages/ember-application/tests/system/visit_test.js:
```diff
@@ -30,59 +73,347 @@ if (isEnabled('ember-application-visit')) {
+        this.route('c');
+      });
+
+      App.visit('/').then((instance) => {
```
`return` the promise

stefanpenner commented on packages/ember-application/tests/system/visit_test.js:
```diff
@@ -90,49 +421,456 @@ if (isEnabled('ember-application-visit')) {
+      });
+    });
+
+    Ember.RSVP.all([a, b, c, d, e]).finally(() => QUnit.start());
```
`return` promise

stefanpenner commented on packages/ember-application/tests/system/visit_test.js:
```diff
@@ -90,49 +421,456 @@ if (isEnabled('ember-application-visit')) {
+      });
     });
+
+    let d = run(function() {
```
seems strange that `visit` requires a run...

chancancode replied:
I don't disagree. I admit that after working on this PR, whatever mental model I thought had about the runloop has been completely destroyed :P

I think I mostly just inherited these patterns from copying the existing tests. But in this particular case I believe removing it would fail the test because the first `visit` here calls `boot`, which has a `run.once("actions", ...);` line somewhere along the way.

chancancode commented on packages/ember-application/lib/system/application-instance.js:
```diff
@@ -212,6 +298,255 @@ let ApplicationInstance = EmberObject.extend(RegistryProxy, ContainerProxy, {
+          throw new Error(error.message);
+        } else {
+          throw error;
+        }
```
I'm not 100% sure this is correct, someone pointed out to me that using `router.router.activeTransition` might be (incorrectly) relying on sync behavior. (Although I see other places using it too.)

However, since the docs/tests is describing the intended behavior (it should follow all redirects and then resolve when rending work is complete, if any), and this implementation is correct as far as the current tests are concerned, if no one has any immediate suggestions this shouldn't be a blocker either. If someone eventually discovered the theoretical bug we can change the implementation to match the intended semantics.

chancancode committed:
Just use the Promise constructor here

Also, don't add the function's internal states into the constructor for
no reason.

chancancode committed:
Use ES6 default parameters

chancancode committed:
rootElement isn't observation: don't need set/get

Also, we don't use _environment on the instance itself, so don't need to
save it off for no reason.

chancancode committed:
Fix all the typos

chancancode committed:
Prefix boolean flags w/Java conventions

chancancode committed:
Use promises for async tests

chancancode committed:
Missed on rename

chancancode committed:
Soften the language, globals mode is still supported

chancancode committed:
Missed a typo

chancancode committed:
Merge branch 'master' into visit_api

Conflicts:
  packages/ember-application/lib/system/application.js
  packages/ember-application/lib/system/application-instance.js
  packages/ember-application/tests/system/visit_test.js
  packages/ember-views/lib/mixins/view_support.js

chancancode committed:
Rollback a few non-critical changes to derisk

The gets the unflagged branches more closer aligned with master; most
important one being `boot` is back to `@private` again since there are
no real reason you need to call it (`visit` calls it automatically), and
because we cannot feature flag docs (and have one of them public, the
other one private).

chancancode committed:
Fix node tests to use new APIs

wycats commented:
The public API here is not final yet, but it's getting close.

It replaces older unstable code behind a flag, so I'm going to merge it so that people who are experimenting with the `visit()` API in Canary can continue to provide feedback, and we can hopefully stabilize this for 2.3.
